src/excmds.ts

```diff
@@ -2098,6 +2098,50 @@ export async function tabprev(increment = 1) {
     })
 }
 
+/**
+ * Pushes a tab to another window. Only works for windows of the same type
+ * (can't push a non-private tab to a private window or a private tab to
+ * a non-private window).
+ */
+//#background
+export async function tabpush(windowId: number) {
+    return activeTabId().then(tabId => browser.tabs.move(tabId, { index: -1, windowId }))
+}
+
+/**
+ * Given a string of the format windowIndex.tabIndex, returns a tuple of
+ * numbers corresponding to the window index and tab index or the current
+ * window and tab if the string doesn't have the right format.
+ */
+//#background_helper
+async function parseWinTabIndex(id: string) {
+    const windows = (await browser.windows.getAll()).map(w => w.id).sort((a, b) => a - b)
+    if (id === null || id === undefined || !id.match(/\d+\.\d+/)) {
+        const tab = await activeTab()
+        const prevId = id
+        id = windows.indexOf(tab.windowId) + "." + (tab.index + 1)
+        logger.info(`taball: Bad tab id: ${prevId}, defaulting to ${id}`)
+    }
+    const [winindex, tabindex_string] = id.split(".")
+    return [windows[parseInt(winindex, 10) - 1], parseInt(tabindex_string, 10) - 1]
+}
+
+/**
+ * Puts a tab identified by a windowIndex.tabIndex id in the current window.
+ * Only works for windows of the same type (can't grab a non-private tab from a
+ * private window and can't grab a private tab from a non-private window).
+ */
+//#background
+export async function tabgrab(id: string) {
```

Comment:
Why is this a string when the other is a number?

Reply:
Because to grab a tab from another window you need to identify both tab and window. Window+String ids are provided by the TabAll completion source which provides strings.